vignettes/hypothesis_testing.Rmd

```diff
@@ -0,0 +1,129 @@
+1. If $A$ and $B$ are both numerical, then $F_{A*B}$ do not contain $F_{A}$ or $F_{B}$.
+1. If $A$ is categorical and $B$ is numerical, then $F_{A*B}$ contains $F_{B}$ but not $F_{A}$.
+
+# Type II Hypothesis Testing
+
+For type II hypothesis testing, the contrasts can be obtained through the following steps:
+
+Create a contrast matrix $L$, with rows equal to the number of parameters associated with the effect,
+columns equal to the number of all parameters.
+
+1. All columns of $L$ associated with effect not containing $F_1$ (except $F_1$) are set to 0.
+1. The submatrix of $L$ associated with $F_1$ is set to $(X_1^\top M X_1)^{-}X_1^\top M X_1$
+1. Each of the remaining submatrices of L associated with an effect $F_2$ that contains $F_1$ is $(X_1^\top M X_1)^{-}X_1^\top M X_2$
+
+where $X_0$ stands for columns of $X$ whose associated effect do not contain $F_1$,
+$X_1$ stands for columns of $X$ associated with $F_1$,
+$X_2$ stands for columns of $X$ whose associated effect contains $F_1$,
+$M = I - X_0(X_0^\top X_0)^{-}X_0^\top$,
+and $Z^{-}$ stands for the g2 inverse of $Z$.
+
+Note: In `mmrm` package, we do not allow singularity in general, so the g2 inverse is the inverse.
+Thus we can use identity matrix in step 2.
+
+# Type III Hypothesis Testing
+
+For type III hypothesis testing, we assume that the hypothesis tested should be the same for all
+designs with the same general form of estimable functions.
+It can be defined with the following step:
+
+Create a contrast matrix $L$, with rows equal to the number of parameters associated with the effect,
+columns equal to the number of all parameters.
+
+1. All columns of $L$ associated with effect not containing $F_1$ (except $F_1$) are set to 0.
+1. The submatrix of $L$ associated with $F_1$ is set to identity matrix.
+1. For each of the remaining submatrices of L associated with an effect $F_2$ that contains $F_1$, equate the
+values.
+
+Here equate the values means that, given that the coefficients in step 2 are all 1 (diagonal), in step 3
+this 1 is divided by the total number of new levels.
+
+# Hypothesis Testing in SAS
+
+In `PROC MIXED` we can specify `HTYPE=1,2,3` to enable these hypothesis testings, and we can use option `e1 e2 e3` to
+view the estimable functions.
+
+## Special Considerations
+
+For `PROC MIXED`, it is important that the categorical values have the correct order, especially
+for structured covariance (otherwise the correlation between visits can be messed up).
+We usually use `CLASS variable(REF=)` to define the reference level.
+However, SAS will put the reference level in the last, so if the visit variable is included in fixed effect,
+there can be issues with the comparison between SAS and R, as in R we still use the first level as our reference.
+
+When we conduct the hypothesis testing, we will also face this issue.
+The result will be slightly different because of different reference groups, however, if correct contrast is used
```

Comment:
but is it possible always to match results by custom options for reference levels? or not?
would be good to make that clear

Reply:
I think it is already clear that `if correct contrast is used we will get closer result`. (I used closer because anyway we will not have identical results)

However, we do not include the algorithm to create the correct reference.

Reply:
Hm it is not clear to me yet :-) Is the way we show the code currently getting us "closest" to SAS results? or is there another way to get even closer?
If there is an algorithm to create the correct reference/contrast then let's include it

Reply:
well this is not "correct reference" but user's choices. e.g. for RACE we really don't need to say "which reference is correct, Aisan or White"? they essentially try to answer the same question, however the result can be a little different. For type II it will be a little complicated to obtain the contrast for the with different reference level, for type III it is easier, e.g. if there are three levels, using lvl1 as reference, we are testing `lvl2 - lvl1` and `lvl3 - lvl1` are both zero, and the contrast should be
```
1 0
0 1
```

however if we are using `lvl3` as reference, then we should test `lvl1 - lvl3`, `lvl2 - lvl3`. However in R we have `lvl2-lvl1`, `lvl3 - lvl1`. With a linear combination of these we can use

```
0 1
-1 -1
```
So here I use "correct contrast" to say that, e.g. in SAS we are testing `lvl1 - lvl3`, `lvl2 - lvl3`, in R we shoud not use identity matrix if we only have `lvl2 - lvl1` and `lvl3 - lvl1` and identity matrix, but use the previous form of matrix to correctly test `lvl1 - lvl3`, `lvl2 - lvl3`. It is not related to "correct reference", but "correct contrast" to reflect the same testing.

However a better solution is to use a different model to create the same design matrix.

In summary, I think we should not provide such functionality to update the contrast afterwards

Reply:
ok makes sense, can you add this example to the vignette? I think that helps the user

Reply:
updated